### ISSUE TITLE
Switch Mautic to be only installable/upgradable via Composer as regular dependency

### ISSUE THREAD

**Issue by PatchRanger** [mautic/mautic#2364]
| Q | A |
| --- | --- |
| Bug report? | n |
| Feature request? | n |
| Enhancement? | y |
## Motivation

For sure all of us have some "hacks" to make Mautic fit their own needs: `.htaccess`, `composer.json`, `index.php`, could be even Mautic source code (until related PRs get merged).
All of these changes are vitally required for the concrete Mautic installation to run.
But all of them would be overridden in case of upgrade.
So upgrade for quite complex installations (well, in fact for "not the simplest") never is done by one click as it is intended to be - it's always connected with re-applying old patches, which is quite annoying and really prevents from upgrading to the most recent version.
While (I guess it is commonly known fact - but nevertheless) encouraging users to use the most recent version very much helps to debug & develop further.
## Proposed solution

What if we consider Mautic as regular composer dependency, which of course is complex enough to have its own dependencies?
I mean something like that:

```
- .htaccess
- composer.json
- index.php
- vendor
-- (other vendors)
-- mautic
-- (other vendors)
```

So:
- `.htaccess` has its own custom logic - and redirects the request to Mautic.
- `composer.json` has a dependency of `mautic/mautic` - and everything else what is needed.
- `index.php` has all of the required stuff - and `require_once`'s original Mautic `index.php`.
  Configuration files could also be moved to document root instead of Mautic directory. It would make possible to divide Mautic-related stuff & Mautic-based-custom-application stuff - so it would look like frameworks are currently installed. We could then create special boilerplate generator, which generates application skeleton with all of the necessary directories & files with meaningful defaults to make it working out of the box - but it's not the case for now.

If, in addition to this, some modifications to Mautic code are needed, then one could use [composer-patches](https://github.com/cweagans/composer-patches) library - which (please note) can't be used in the current situation.

Also one could simply test Mautic PR: just add a line to `composer.json` (referring to corresponding patch, using knowledge that all of Github PRs are available as diff: http://stackoverflow.com/a/6188624 ) and run `composer update`.

So all of the custom changes are:
- Declarative: you always know what and where was changed.
- Part of upgrade workflow: they are "executable documentation" of all the hacks made, so you never forget to update the patch or accidentally override important hack - as if you do upgrade, all of the "hacks" are re-applied, automagically.

It also has additional benefit: we could convert our current install & upgrade scripts into `post-install-cmd` & `post-update-cmd` composer commands. It could be as simple as `doctrine:migrations:migrate` - but is not limited to such simple as we could put there our own upgrade scripts. Then one-click-upgrade UI should just rely on the same upgrade progress: just run `composer update` on click.
## TL;DR Benefits
- **Central storage (via [composer-patches](https://github.com/cweagans/composer-patches)) of all custom modifications, which get re-applied after each update**: Encourages users to upgrade sooner and more often.
- **Easier adoption & testing of PRs**: It would be extremely easier to apply PR, which would result in increasing development speed of Mautic.
- **Code & database upgrades would be tied together**: No one could get the recent code without applying the corresponding database updates. It would decrease the possibility of issues as we are getting the tool, which can not be used incorrectly.
- **Both UI & CLI rely on the same mechanism**: Increasing code reusage, which is good.
